### PR TITLE
Add async docs

### PR DESCRIPTION
**Pull request by apurvam**

### CODE REVIEW

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:10]: I think we should say 0.25 -- I'm just waiting on one small fix for something the fault injection caught in soak and then I'm going to cut a new release, so it should be out before these docs go live

File: docs/reference/async-processing.mdx
Comment (agavra) [inline docs/reference/async-processing.mdx:47]: @ableegoldman can we wrap our async processor in a wrapper that would allow you to turn it totally off via config? basically if you set the num threads to 1 it doesn't use any of the async mechanisms and just defaults back to the basic behavior

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:18]: I know this was my own wording, but I'm wondering if we should say "large enough" keyspace, or hedge it even further? I'm worried this sounds like it won't help at all unless you have a huge number of unique keys, which isn't exactly true -- to be fair I haven't tested very much with smaller keyspaces, but it should still improve things as long as you have eg more keys than async threads, which should always be the case and would be a very small keyspace indeed

Does that make sense? Not exactly sure how to reword this..I guess we could just remove the "with a large keyspace" part of the sentence altogether...thoughts?

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:20]: Have we settled on "Responsive Async Processor" as a title?

(not saying I don't like the term, just wondering if we've actively decided to commit to this. If not, we probably shouldn't capitalize it -- or should just figure out what we want to call it. FWIW I do think "Responsive Async Processor" is fine, as long as everyone else likes it)

File: docs/reference/async-processing.mdx
Comment (agavra) [inline docs/reference/async-processing.mdx:103]: obviously out of scope for this PR, but why didn't we choose to make the `async.thread.pool.size` a static number? seems like having one config depending on another config is quite confusing (@ableegoldman )

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:24]: Maybe "which makes a remote or long-blocking call"? Technically people can have very long running processors that don't interact with remote systems -- though that's rare. But is it only rare because, like RPCs, it doesn't work very well with regular Kafka Streams? 🤔

File: docs/reference/async-processing.mdx
Comment (agavra) [reply docs/reference/async-processing.mdx:82]: ah I just saw the note in limitations.

**Review by agavra — APPROVED** [review]
😆 left comments for @apurvam on the vercel page and comments for @ableegoldman in this PR on GitHub. LGTM though!

File: docs/reference/async-processing.mdx
Comment (agavra) [reply docs/reference/async-processing.mdx:20]: I think we should drop "Responsive" if it's on the Responsive site since in that context it's obvious that it's Responsive's processor. (c.f. https://brand.hashicorp.com/product_logos)

File: docs/reference/async-processing.mdx
Comment (apurvam) [reply docs/reference/async-processing.mdx:18]: Yes. Let's drop it and focus on the use case it was built for. I'll reword.

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:28]: thanks for putting together this graph -- it really shows how performance is freed from the influence of long blocking calls. I'm sure the async throughput in the 1ms case would actually be even higher except we were running up against the limits of that node/the resource limits. Still a great way to showcase async performance -- if anything it really drives home how async will let you scale to make use of all the resources available and become limited only by hardware, rather than RPC latency!

Comment (ableegoldman) [reply docs/reference/async-processing.mdx:28]: Actually we should say pretty much exactly that -- we sort of hint at it right now but it would be great to really drive home the point that async processing frees up your application to utilize all the available resources, so you can truly scale up your compute independently! 

..or something like that

File: docs/reference/async-processing.mdx
Comment (apurvam) [reply docs/reference/async-processing.mdx:20]: Agreed. I'll drop the 'Responsive' prefix.

File: docs/reference/async-processing.mdx
Comment (apurvam) [reply docs/reference/async-processing.mdx:24]: good call. will update.

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:35]: Maybe here is where it would be good to point out what I was getting at above -- the lack of change in async throughput means we're only limited by the resources available to this client. With better hardware, the async processor would likely be pushing even higher traffic rates!

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:34]: One nit about the graphic and how we talk about it here -- it feels a little weird to have 2ms first and 1ms last. I know it makes sense for us since the lower latency actually means higher throughput, but it just looks a little weird to see the x axis moving backwards on the graph. took me an extra second to understand what I was looking at.

not a huge deal though, your call, just wanted to mention it since if it bothers/confused, me it'll probably bother other people too 🤷‍♀️

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx]: This is a great line -- your rewording of the original makes it really hit hard. Nice editing 🙂

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [reply docs/reference/async-processing.mdx:47]: that's a good idea -- yeah it shouldn't be too hard (famous last words). I'll give it a shot now since we're waiting on one more fix before cutting the new release

File: docs/reference/async-processing.mdx
Comment (apurvam) [reply docs/reference/async-processing.mdx:34]: I think you are right. I went back and forth over it. I think it's more conventional to have the x-axis be ascending. Will update.

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [reply docs/reference/async-processing.mdx:47]: ok it's actually not simple at all haha 😭  -- the usual culprit of "configs aren't available at topology build time". It's still possible just not as easy as I was hoping -- I'll timebox this and file a ticket to follow up on if I can't get it working in ~30 minutes or so

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:82]: ah I just double-checked and we have an earlier check for this that  throws a better exception:

```suggestion
You will know if this is not set up correctly by an immediate `UnsupportedOperationException` upon startup.
```

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [reply docs/reference/async-processing.mdx:81]: yeah it's just a silly artificial requirement due to how we connect things internally right now --we can and absolutely should fix this in the future, it's just non-trivial so it was put on the backburner since this feature is most sorely needed by stateful apps anyway. But good question (should we maybe say something here too, or add a note to see the "Limitations" section for an explanation on this?

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [reply docs/reference/async-processing.mdx:117]: I guess I personally found it easier to reason about in terms of "number of threads per StreamThread"

Also, that way it's always exact, vs if we make it a static number (I assume you mean like a config for the actual total number of async threads?) then the actual number of threads may not match if its not an exact multiple of the `num.stream.threads`

Mainly though I figured it like this: if someone wants to upp the number of StreamThreads that probably means they have more capacity and would want to increase the number of async threads with it, to keep the async threads per StreamThread a constant. That was just my reasoning at the time -- I can see good arguments for doing it as a "static" config. WDYT @agavra  (cc @rodesai )

Comment (agavra) [reply docs/reference/async-processing.mdx:117]: > then the actual number of threads may not match
I imagined it would be a shared pool across all threads so you can easily reason about the concurrency to the remote system

File: docs/reference/async-processing.mdx
Comment (agavra) [reply docs/reference/async-processing.mdx:46]: Without this it may actually be harder for customers to test/bake in a staging environment. I know every customer has per-env configs but most don't have per-env code deployments.

Comment (apurvam) [reply docs/reference/async-processing.mdx:46]: Another option here would be to set the async threads to 1 and then make the async processor a no-op.

File: docs/reference/async-processing.mdx
Comment (apurvam) [reply docs/reference/async-processing.mdx:117]: This is a tricky one. 

I think we may want to adjust based on feedback, but if one of the concerns is the managing capacity of the external system, then having it capped statically makes sense. Basically the question is 'what is the max parallelism' that the external system can tolerate?'. Then we set the num async threads based on that. It isn't perfect though, because more Kafka Streams nodes means more threads and more parallelism anyway. 

So it probably doesn't matter that much, and we can adjust based on feedback.

Comment (apurvam) [reply docs/reference/async-processing.mdx:117]: In other words, in either approach the parallelism to the external system is variable _and_ a deterministic function of the number of nodes.

File: docs/reference/async-processing.mdx
Comment (apurvam) [reply docs/reference/async-processing.mdx:81]: Got it. will update.

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:26]:
```suggestion
    alt={"Responsive Async Processor Peformance"}
```

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:68]:
```suggestion
        createAsyncProcessorSupplier(new MyStatefulProcessorSupplier<>()),
        Named.as("AsyncProcessor"),
        STORE_NAME);
```

File: docs/reference/async-processing.mdx
Comment (ableegoldman) [inline docs/reference/async-processing.mdx:37]: > The async processor makes high round trip network latency a non-factor in achieving high throughput

I feel like this is such a good line, and putting it in the warning kind of lessens its impact. On the other hand it also lessens the downside of what's in the warning so it does its job. 

Not saying you need to change anything here -- I guess what I'm saying is this would be a good tagline for async processing (like maybe when we do the blog post cc @agavra )

**Pull request merged**
